### DISCUSSION TITLE
Highlighting and discussing code lines at once

### DISCUSSION THREAD

**Discussion by reuning** [quarto-dev/quarto-cli#4022] [Q&A]
Using RevalJS I was wondering if it is possible to step through line highlights and provide commentary for each line. Below is an example of my attempt to do this, but it doesn't work out the way I want--the non-code fragments all appear in between changing line highlights. 

I added slashes so that it wouldn't work 

````
## My Solution 

```{.R code-line-numbers="2|3"}
iris %>% 
    drop_na(Species) %>% 
    group_by(Species) %>%
    summarize(n=n(), mean=mean(Sepal.Width, na.rm=T),
            median=median(Sepal.Width, na.rm=T))
```

::: {.fragment .fade-out fragment-index=1}
Filtering out missing values
:::

::: {.fragment fragment-index=1}
Grouping by missing values
:::
````

**Comment by cscheid** [comment]
(wrap your overall example with _more_ backticks than used in the example itself and it works.)

**Reply by reuning** [reply]
Thanks. I swear I did that and couldn't figure out what was going on but clearly forgot them.

**Comment by cderv** [comment]
Stepping through code and fragment are both reveal.js native feature. Quarto only handles the support and improved markdown syntax. 

It seems those two features don't work well together and I don't know if something like is possible in reveal.js. 

Maybe using reveal.js hooks, something can be built in JS 🤔 

One trail to follow would be to look into reveal.js community if someone has made something like that working. If so we could support, otherwise it would need to be develop. 

Thanks for the idea!

**Reply by reuning** [reply]
For what it is worth, I found this: https://github.com/bnjmnt4n/reveal-code-focus 

It seems to do what I'm talking about. Hope to see this in the future.

**Reply by shafayetShafee** [reply]
@reuning , what about writing a quarto filter extension to do this? 😉

**Reply by reuning** [reply]
I should be prepping for the beginning of the semester... but in someways wouldn't this be prepping?

**Reply by shafayetShafee** [reply]
YESS!!, I am in the same position but got my whole interest in this plugin for now 😂

**Reply by cderv** [reply]
This is a really cool plugin ! I would love to see that as a Quarto extension ! It is not among known RevealJs quarto extensions I believe (from this list https://github.com/mcanouil/awesome-quarto#extensions)

Do not hesitate to ping me for any help if needed. 😉

**Reply by shafayetShafee** [reply]
@cderv, I need your help actually. I have implemented the plugin as a custom format [`codefocus`](https://github.com/shafayetShafee/codefocus) successfully but I am not sure how to implement this as a plugin as described in [quarto doc](https://quarto.org/docs/extensions/revealjs.html#example-attribution). I have tried to give a detailed description of my problem and so far what I did in the [repo readme](https://github.com/shafayetShafee/codefocus#need-help).

Thanks 😊🙏

**Reply by cderv** [reply]
great. I'll have look as soon as possible

**Reply by shafayetShafee** [reply]
Thank you so much 😊

**Comment by reuning** [comment]
So I put together a plugin here. https://github.com/reuning/codefocus

Lord knows that it has been 15 years since I took a college course on JavaScript. But it seems to work, it uses the line-highlight css to try to keep some consistency. I also checked that it doesn't seem to cause any problem with using line-highlight in other parts of your presentation. Any thoughts/feedback welcome. Though I should probably but this down for a bit.

**Reply by reuning** [reply]
Also fyi @shafayetShafee

**Reply by shafayetShafee** [reply]
Woah, Great. This extension completely supersedes the [`codefocus-revealjs`](https://github.com/shafayetShafee/codefocus) format that I have gathered. Thanks. By the way, I would recommend putting a demo (by activating GitHub page) in your repo.

**Reply by reuning** [reply]
And [done](https://reuning.github.io/codefocus/example.html).

**Reply by reuning** [reply]
@cderv if you have a moment, I'm having trouble installing it off of github. When I run `quarto add reuning/codefocus` I get `Extension not found in local or remote sources`. I've tried also `quarto install extension reuning/codefocus` and get the same issue. 

I've tried this from my macbook and windows computer. I can clone the github repo fine. I feel like there is something silly I'm missing here.

**Reply by dragonstyle** [reply]
It looks to me like the issue is that we're expecting a branch named `main` and not handling it super gracefully. I think this should be pretty straightforward as Github provides UI for this now.

Would renaming the default branch be possible for you?

**Reply by reuning** [reply]
That fixed it. Thank you.

**Reply by dragonstyle** [reply]
Awesome! I'm not sure if there is a better way for us to discover the default branch name so for the time being we just always assume `main`. Glad this got things up and running!

**Reply by cderv** [reply]
@reuning and @shafayetShafee pretty quick on project! Nicely done. Sorry I did not have time to look at it yet as onto something right now, but it seems it is sorted out and [reuning/codefocus](https://github.com/reuning/codefocus) is the extension for this. 

Great to have such community with us ! Thank you!
